internal/driver/driver.go

```diff
@@ -0,0 +1,793 @@
+	for i, req := range reqs {
+		s.lc.Debugf("S7Driver.HandleWriteCommands: protocols: %v, resource: %v, parameters: %v, attributes: %v", protocols, req.DeviceResourceName, params[i], req.Attributes)
+
+		variable := cast.ToString(req.Attributes["NodeName"])
+		dbInfo, _ := s.getDBInfo(variable)
+
+		reading, err := newCommandValue(req.Type, params[i])
+		if err != nil {
+			s.lc.Errorf("newCommandValue error: %s", err)
+		}
+		helper.SetValueAt(datas[i], 0, reading)
+
+		// create gos7 DataItem
+		var item = gos7.S7DataItem{
+			Area:     dbInfo.Area,
+			WordLen:  dbInfo.WordLength,
+			DBNumber: dbInfo.DBNumber,
+			Start:    dbInfo.Start,
+			Amount:   dbInfo.Amount,
+			Data:     datas[i],
+			Error:    errors[i],
+		}
+		items = append(items, item)
+
+	}
+	fmt.Println("S7DataItems: ", items)
+
+	// send command requests
+	times := int(reqs_len / batch_size)
+	remains := reqs_len % batch_size
+	var reqs1 []sdkModel.CommandRequest
+	var items1 = []gos7.S7DataItem{}
+	s7Client := s.getS7Client(deviceName)
+
+outloop:
+	for j := 0; j <= times; j++ {
+		items1 = items1[:0]
+		reqs1 = reqs1[:0] // clear array
+		if j >= times {
+			if remains > 0 {
+				items1 = items[times*batch_size : times*batch_size+remains]
+			} else {
+				break outloop // load complete, exit loop
+			}
+		} else {
+			items1 = items[j*batch_size : j*batch_size+batch_size]
+		}
+		// fmt.Println("reqs1: ", reqs1, "items1: ", items1)
+
+		// write data to S7
+		// fmt.Println("reqs len: ", len(items))
+		err = s7Client.Client.AGWriteMulti(items1, len(items1))
+		if err != nil {
+			s.lc.Errorf("AGWriteMulti Error: %s, reconnecting...", err)
+			s.s7Clients[deviceName] = s.NewS7Client(deviceName, protocols)
+			return err
+		} else {
+			//fmt.Println(datas)
+		}
+	}
+	for _, err1 := range errors {
+		if err1 != "" {
+			s.lc.Errorf("S7 Client AGWriteMulti error: %s", err1)
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Stop the protocol-specific DS code to shutdown gracefully, or
+// if the force parameter is 'true', immediately. The driver is responsible
+// for closing any in-use channels, including the channel used to send async
+// readings (if supported).
+func (s *Driver) Stop(force bool) error {
+	// Then Logging Client might not be initialized
+	if s.lc != nil {
+		s.lc.Debugf("Driver.Stop called: force=%v", force)
+	}
+	return nil
+}
+
+// AddDevice is a callback function that is invoked
+// when a new Device associated with this Device Service is added
+func (s *Driver) AddDevice(deviceName string, protocols map[string]models.ProtocolProperties, adminState models.AdminState) error {
+	s.lc.Debugf("a new Device is added: %s", deviceName)
+	return nil
+}
+
+// UpdateDevice is a callback function that is invoked
+// when a Device associated with this Device Service is updated
+func (s *Driver) UpdateDevice(deviceName string, protocols map[string]models.ProtocolProperties, adminState models.AdminState) error {
+	s.lc.Debugf("Device %s is updated", deviceName)
```

Comment:
Should determine if the S7 Client needs to be recreated since the device changed.

Reply:
That is dependent on you S7 protocol. If one of these values changed...
```
        Host: 192.168.123.199
        Port: 102
        Rack: 0
        Slot: 1
```
You can look up the current device from the cache and compare the protocol properties to the new one passed in.